vassal-app/src/main/java/VASSAL/build/module/index/IndexManager.java

```diff
@@ -0,0 +1,177 @@
+/*
+ *
+ * Copyright (c) 2023 by The VASSAL Development Team
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Library General Public
+ * License (LGPL) as published by the Free Software Foundation.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU Library General Public
+ * License along with this library; if not, copies are available
+ * at http://www.opensource.org.
+ */
+package VASSAL.build.module.index;
+
+import VASSAL.build.GameModule;
+import VASSAL.build.module.Chatter;
+import VASSAL.build.module.Map;
+import VASSAL.counters.GamePiece;
+import VASSAL.counters.Stack;
+
+import java.awt.Point;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+/**
+ * The IndexManager maintains a set of indexes on GamePieces that will primarily be used by
+ * the GKC Fast-match mechanism to quickly find sets of pieces that meet certain critera.
+ *
+ * It maintains the following indexes:
+ *  o A per-map quadtree of piece locations
+ *  o A per-map cross-reference of values of the Currentzone and LocationName properties
+ *
+ *  The indexes for each Map are stored in a VassalMapPieceIndex object
+ *
+ *  NOTE: Pieces not on Maps are NOT included in any index.
+ */
+public class IndexManager {
+
+  /**
+   * HashMap of the indexes created so far
+   */
+  private final java.util.Map<Map, VassalMapPieceIndex> indexes = new HashMap<>();
+
+  /**
+   * Return the combined index object for the specified Map, create one if needed
+   *
+   * @param map Map
+   * @return Index
+   */
+  private VassalMapPieceIndex getIndex(Map map) {
+    VassalMapPieceIndex index = indexes.get(map);
+    if (index == null) {
+      index = new VassalMapPieceIndex(map);
+      indexes.put(map, index);
+    }
+    return index;
+  }
```

Comment:
You can do this whole function as a one-liner:
```
return indexes.computeIfAbsent(map, m -> new VassalMapPieceIndex(m));
```

Reply:
Done